cpp/src/lance/io/record_batch_reader.cc

```diff
@@ -70,19 +80,44 @@ std::shared_ptr<::arrow::Schema> RecordBatchReader::schema() const {
 
 ::arrow::Status RecordBatchReader::ReadNext(std::shared_ptr<::arrow::RecordBatch>* batch) {
   int32_t batch_id = current_batch_++;
-  if (batch_id < reader_->metadata().num_batches()) {
-    ARROW_ASSIGN_OR_RAISE(auto batch_read, project_->Execute(reader_, batch_id));
-    if (batch_read) {
-      *batch = std::move(batch_read);
-    }
+  ARROW_ASSIGN_OR_RAISE(auto batch_read, ReadBatch(batch_id));
+  if (batch_read) {
+    *batch = std::move(batch_read);
   }
   return ::arrow::Status::OK();
 }
 
+::arrow::Result<std::shared_ptr<::arrow::RecordBatch>> RecordBatchReader::ReadBatch(
+    int32_t batch_id) const {
+  if (batch_id < reader_->metadata().num_batches()) {
+    return project_->Execute(reader_, batch_id);
+  }
+  return nullptr;
+}
+
 ::arrow::Future<std::shared_ptr<::arrow::RecordBatch>> RecordBatchReader::operator()() {
-  /// TODO: Make it truly async someday.
-  auto f = ::arrow::Future<std::shared_ptr<::arrow::RecordBatch>>::MakeFinished(this->Next());
-  return f;
+  int total_batches = reader_->metadata().num_batches();
+  while (static_cast<int32_t>(readahead_queue_.size()) < options_->batch_readahead &&
+         current_batch_ < total_batches) {
+    int32_t batch_id = current_batch_++;
+    auto result = thread_pool_->Submit(
+        [&](int32_t batch_id) {
+          /// TODO: how to handle error in thread pool?
+          auto result = this->ReadBatch(batch_id);
+          return result.ValueOrDie();
```

Comment:
I forgot: does this override an Arrow function? Could we return `Future<Result<shared_ptr<RecordBatch>>>`?

Reply:
The ThreadPool.submit() seems not accept a `Result<T>`, you mean directly return `Future<Result<T>>` within each thread? I can take a look whether it works tho.

Reply:
Yea,  returns `Future<Result<T>>` works. Thanks for the suggestion. 

Updated the code accordingly.